x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CreateAndFollowIndexAction.java

```diff
@@ -0,0 +1,309 @@
+                assert entry.getValue().size() == 1;
+                String clusterNameAlias = entry.getKey();
+                String leaderIndex = entry.getValue().get(0);
+
+                Client remoteClient = client.getRemoteClusterClient(clusterNameAlias);
+                ClusterStateRequest clusterStateRequest = new ClusterStateRequest();
+                clusterStateRequest.clear();
+                clusterStateRequest.metaData(true);
+                clusterStateRequest.indices(leaderIndex);
+                remoteClient.admin().cluster().state(clusterStateRequest, ActionListener.wrap(r -> {
+                    ClusterState remoteClusterState = r.getState();
+                    IndexMetaData leaderIndexMetadata = remoteClusterState.getMetaData().index(leaderIndex);
+                    createFollowIndex(leaderIndexMetadata, request, listener);
+                }, listener::onFailure));
+            }
+        }
+
+        private void createFollowIndex(IndexMetaData leaderIndexMetaData, Request request, ActionListener<Response> listener) {
+            if (leaderIndexMetaData == null) {
+                listener.onFailure(new IllegalArgumentException("leader index [" + request.getFollowRequest().getLeaderIndex() +
+                    "] does not exist"));
+                return;
+            }
+
+            ActionListener<Boolean> handler = ActionListener.wrap(
+                result -> {
+                    if (result) {
+                        initiateFollowing(request, listener);
+                    } else {
+                        listener.onResponse(new Response(true, false, false));
+                    }
+                },
+                listener::onFailure);
+            clusterService.submitStateUpdateTask("follow_index_action", new AckedClusterStateUpdateTask<Boolean>(request, handler) {
+
+                @Override
+                protected Boolean newResponse(boolean acknowledged) {
+                    return acknowledged;
+                }
+
+                @Override
+                public ClusterState execute(ClusterState currentState) throws Exception {
+                    IndexMetaData currentIndex = currentState.metaData().index(request.getFollowRequest().getFollowIndex());
+                    if (currentIndex != null) {
+                        throw new ResourceAlreadyExistsException(currentIndex.getIndex());
+                    }
+
+                    MetaData.Builder mdBuilder = MetaData.builder(currentState.metaData());
+                    if (request.getFollowRequest().getFollowIndex() != null) {
+                        IndexMetaData.Builder imdBuilder = IndexMetaData.builder(leaderIndexMetaData);
+
+                        Settings.Builder settingsBuilder = Settings.builder();
+                        settingsBuilder.put(leaderIndexMetaData.getSettings());
+                        // TODO: do we want leader and follow index to have the same UUID?
```

Comment:
As you note one line below, we can not do this if the indices are intra-cluster yet I am not sure if I see any advantage to doing this if the indices are inter-cluster.

Reply:
Thanks for checking, I just wanted to point this out during the review. I will remove the todo.